File: 프로그래머스/3/43162. 네트워크/네트워크.py
def solution(n, computers):
    # 방문 여부 체크
    visited = [False for _ in range(n)]
    
    # 연결 여부 초기화
    graph = [[False for _ in range(n)] for __ in range(n)]
    for i in range(n):
        for j in range(n):
            if computers[i][j] == 1:
                graph[i][j] = True
                
    queue = []
    answer = 0
    
    for i in range(n):
        if visited[i]:
            continue
        # print(i)
        answer += 1
        queue.append(i)
        visited[i] = True
        bfs(visited, graph, queue)
        # print(visited, graph, queue)
    
    
    return answer


def bfs(visited, graph, queue):
    n = len(graph)
    while queue:
        cur = queue.pop()
        
        for i in range(n):
            if not graph[cur][i] or visited[i]:
                continue
            queue.append(i)
            visited[i] = True
            

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] computers) {
        boolean[] checked = new boolean[n];
        ArrayList<Integer>[] graph = new ArrayList[n];
        for(int i = 0; i < n; i++){
            graph[i] = new ArrayList<>();
        }
        
        for(int i = 0; i < n; i++){
            for(int j = i + 1; j < n; j++){
                if(i == j) continue;
                if(computers[i][j] == 1){
                    graph[i].add(j);
                    graph[j].add(i);
                }
            }
        }
        
        // System.out.println(Arrays.toString(graph));
        
        int answer = 0;
        
        for(int i = 0; i < n; i++){
            if(checked[i]) continue;
            bfs(i, graph, checked);
            answer++;
        }
        return answer;
    }
    
    public void bfs(int start, ArrayList<Integer>[] graph, boolean[] checked){
        Deque<Integer> queue = new ArrayDeque<>();
        checked[start] = true;
        queue.add(start);
        
        while(!queue.isEmpty()){
            int cur = queue.poll();
            
            for(int next:graph[cur]){
                if(checked[next]) continue;
                queue.add(next);
                checked[next] = true;
            }
        }
    }
}